File: solutions/Arrays/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> nge=new Stack<>();
        Map<Integer,Integer> ok=new HashMap<>();
        int[] result=new int[nums1.length];
        int[] c=nums2;
        for(int i=nums2.length-1;i>=0;i--)
        {
            int temp2=nums2[i];

            if(nge.isEmpty())
            {
                nums2[i]=-1;
                ok.put(temp2,-1);
                nge.push(temp2);
                
            }
            else if(nge.peek()<=temp2)
            {
                while(!nge.isEmpty() && nge.peek()<=temp2)
                {
                    nge.pop();
                }
                if (nge.isEmpty())
                {
                    nums2[i]=-1;
                    ok.put(temp2,-1);
                    nge.push(temp2);
                
                    continue;
                }
                nums2[i]=nge.peek();
                ok.put(temp2,nge.peek());
                nge.push(temp2);
                
                
            }
            else
            {
                nums2[i]=nge.peek();
                ok.put(temp2,nge.peek());
                nge.push(temp2);
            }

        }
    for(int i=0;i<nums1.length;i++)
    {
        result[i]=ok.get(nums1[i]);
    }



    return result;

    }
}

File: solutions/Arrays/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack=[]
        index_map= {n:-1 for n in nums1}
        
        for i in range(len(nums2)):
            curr=nums2[i]
            while stack and curr>stack[-1]:
                val=stack.pop()
                index_map[val]=nums2[i]
            if curr in index_map:
                stack.append(curr)
        return [i for i in index_map.values()]
